cai2hcl/common.go

```diff
@@ -2,36 +2,36 @@ package cai2hcl
 
 import (
 	"github.com/GoogleCloudPlatform/terraform-google-conversion/v2/caiasset"
-	tpg "github.com/hashicorp/terraform-provider-google/google"
+	tfschema "github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/zclconf/go-cty/cty"
 )
 
-// Converter interface for resources.
-type Converter interface {
-	// Convert turns assets into hcl blocks.
-	Convert(asset []*caiasset.Asset) ([]*HCLResourceBlock, error)
+// Shared converters context.
+type ConverterContext struct {
+	name   string
+	schema map[string]*tfschema.Schema
 }
 
+// Convert assets into hcl blocks.
+type Converter func(asset []*caiasset.Asset, context *ConverterContext) ([]*HCLResourceBlock, error)
+
 // HCLResourceBlock identifies the HCL block's labels and content.
 type HCLResourceBlock struct {
 	Labels []string
 	Value  cty.Value
 }
 
-// converterNames map key is the CAI Asset type, value is the TF resource name.
+// Map key is the CAI Asset type, value is the TF resource name.
 var converterNames = map[string]string{
-	ComputeInstanceAssetType:       "google_compute_instance",
-	ComputeForwardingRuleAssetType: "google_compute_forwarding_rule",
-	ProjectAssetType:               "google_project",
-	ProjectBillingAssetType:        "google_project",
+	"compute.googleapis.com/Instance":                "google_compute_instance",
+	"compute.googleapis.com/ForwardingRule":          "google_compute_forwarding_rule",
+	"cloudresourcemanager.googleapis.com/Project":    "google_project",
+	"cloudbilling.googleapis.com/ProjectBillingInfo": "google_project",
 }
 
-// converterMap initializes converters by their TF resource name.
+// Map key is the TF resource name, value - conversion function.
```

Comment:
what's the benefit to switch from interface to function?

Reply:
since all resources conversion is in the same package cai2hcl, similar functions like convertIAM will run into naming issues if using functions.

Reply:
Main purpose of PR was to reduce the code amount in each converter and concentrate resource type names (compute.googleapis.com, Instance and google_compute_instance) in a single file.

Yours is a good point: I did not realise that. Let me try to return to instance based converters while keeping the benefits above.

Reply:
Talked to Stephen and he discourages the idea of adding more manual support. So let me cancel this PR for now - will spend some time to investigate possibility of generated converters.